verifiable/onchain/converter.go

```diff
@@ -0,0 +1,474 @@
+package onchain
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	core "github.com/iden3/go-iden3-core/v2"
+	"github.com/iden3/go-iden3-core/v2/w3c"
+	"github.com/iden3/go-merkletree-sql/v2"
+	"github.com/iden3/go-schema-processor/v2/merklize"
+	"github.com/iden3/go-schema-processor/v2/verifiable"
+	"github.com/piprate/json-gold/ld"
+)
+
+type Convertor struct {
+	marklizeInstance merklize.Options
+
+	id string
+}
+
+type ConvertorOption func(*Convertor)
+
+// WithID sets custom id for the credential.
+func WithID(id string) ConvertorOption {
+	return func(p *Convertor) {
+		p.id = id
+	}
+}
+
+// WithMerklizerOptions sets options for merklizer.
+func WithMerklizerOptions(m merklize.Options) ConvertorOption {
+	return func(p *Convertor) {
+		p.marklizeInstance = m
+	}
+}
+
+func NewParser(opts ...ConvertorOption) *Convertor {
+	p := &Convertor{}
+
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	return p
+}
+
+// ConvertVerifiableCredential converts an onchain verifiable credential to a W3C verifiable credential.
+// The w3c credential id will be regenerated.
+func ConvertVerifiableCredential(credential *Credential) (*verifiable.W3CCredential, error) {
+	id := fmt.Sprintf("urn:uuid:%s", uuid.NewString())
```

Comment:
why id of credential is not comming from the contract itself. It is not supposed to be different each time.

Reply:
yes, we should have unique, but still deterministic id. E.g. based on DID of issuer (or chainId+contractAddr) + credential id from contract.

Reply:
@vmidyllic since we are using int as ID in on-chain issuer. Credentials with `id: 1` will not be merklized. We should have valid uri.

Reply:
id must have string format, valid urn, it should come as it from contract. `urn:uiid:1`